README.md
# HttpHeaders::Utils

[![Build Status: master](https://travis-ci.com/XPBytes/http_headers-utils.svg)](https://travis-ci.com/XPBytes/http_headers-utils)
[![Gem Version](https://badge.fury.io/rb/http_headers-utils.svg)](https://badge.fury.io/rb/http_headers-utils)
[![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)

:nut_and_bolt: Utility belt for the HttpHeader libraries

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'http_headers-utils'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install http_headers-utils

## Usage

### List

There is a utility class for list based headers. The utility parses the header values(s) for you and 
outputs a custom entry class. This can be used for almost any header that allows multiple values, as
per the RFC these are consistent with the exception of the Cookie / Set-Cookie headers (at time of 
writing).

```ruby
require 'http_headers/utils/list'

class ListBasedHeader < Utils::List
  def initialize(value)
    super value, entry_klazz: Entry
  end
  
  class Entry
    def initialize(value, index:, parameters:)
    
    end
  end
end
```

## Related

- [HttpHeaders::Accept](https://github.com/XPBytes/http_headers-accept): :nut_and_bolt: Utility to parse and sort the "Accept" HTTP Header
- [HttpHeaders::AcceptLanguage](https://github.com/XPBytes/http_headers-accept_language): :nut_and_bolt: Utility to parse and sort the "Accept-Language" HTTP Header
- [HttpHeaders::ContentType](https://github.com/XPBytes/http_headers-content_type): :nut_and_bolt: Utility to parse and sort the "Content-Type" HTTP Header
- [HttpHeaders::Link](https://github.com/XPBytes/http_headers-link): :nut_and_bolt: Utility to parse and sort the "Link" HTTP Header

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can
also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the
version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version,
push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at [XPBytes/http_headers-utils](https://github.com/XPBytes/http_headers-utils).


CHANGELOG.md
# Changelog

## 0.2.0

- Fix whitespace on input by using `strip`
- Change `List` from `class` to `module`
- Change `List.new` to automatically sort
- Add `Single` for a single header value
- Add `to_header` lookup and implementation for `Single` and `List`
- Change `to_header` to use `HEADER_DELIMITER`

## 0.1.1

- Add a missing require to `delegate`

## 0.1.0

:baby: Initial Release


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "http_headers/utils"

require "minitest/autorun"


test/http_headers/utils_test.rb
require "test_helper"

class HttpHeaders::UtilsTest < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::HttpHeaders::Utils::VERSION
  end
end


test/http_headers/utils/list_test.rb
require 'test_helper'
require 'delegate'

module HttpHeaders
  module Utils
    class ListTest < Minitest::Test
      class TestAccept < DelegateClass(Array)

        def initialize(value)
          list = HttpHeaders::Utils::List.new(value, entry_klazz: Entry)
          __setobj__ list
        end

        def to_s
          HttpHeaders::Utils::List.to_header(self)
        end

        class Entry
          attr_reader :media_type, :index

          def initialize(media_type, index:, parameters:)
            self.media_type = media_type
            self.index = index
            self.parameters = parameters
          end

          def q
            parameters.fetch(:q) { 1.0 }.to_f
          end

          def <=>(other)
            quality = other.q <=> q
            return quality unless quality.zero?
            index <=> other.index
          end

          def to_header
            [media_type].concat(parameters.map { |k, v| "#{k}=#{v}" }).compact.reject(&:empty?).join('; ')
          end

          def to_s
            to_header
          end

          def [](index)
            parameters.fetch(index)
          end

          private

          attr_writer :media_type, :index
          attr_accessor :parameters
        end
      end

      def test_it_parses_empty
        TestAccept.new('')
        pass 'did not break'
      end

      def test_it_parses_one
        list = TestAccept.new('application/json')
        assert_equal 1, list.length
        assert_equal 'application/json', list.first.to_s
      end

      def test_it_parses_parameters
        list = TestAccept.new('application/json; foo=bar; q=0.5')
        assert_equal 1, list.length
        assert_equal 'bar', list.first[:foo]
        assert_equal 0.5, list.first.q
      end

      def test_it_parses_multiple_lines
        list = TestAccept.new(['*/*; q=0.1', 'application/json, text/html; q=0.8'])
        assert_equal 3, list.length
        assert_equal 'application/json', list.first.to_s
        assert_equal '*/*; q=0.1', list.last.to_s
        assert_equal 0.8, list[1].q
      end

      def test_it_can_convert_to_a_header
        list = TestAccept.new(['*/*; q=0.1', 'application/json, text/html; q=0.8'])
        assert_equal 'application/json, text/html; q=0.8, */*; q=0.1', list.to_s
      end
    end
  end
end


lib/http_headers/utils/list.rb
module HttpHeaders
  module Utils

    ##
    # @example Accept values
    #
    #   class Accept < DelegateClass(Array)
    #     def initialize(value)
    #       super HttpHeaders::List.new(value, entry_klazz: Accept::Entry)
    #     end
    #
    #     class Entry
    #       def initialize(media_type, index: parameters:)
    #         ...
    #       end
    #
    #       def q
    #         parameters.fetch(:q) { 1.0 }.to_f
    #       end
    #
    #       def <=>(other)
    #         quality = other.q <=> q
    #         return quality unless quality.zero?
    #         index <=> other.index
    #       end
    #     end
    #   end
    #
    #   Accept.new(['*/*; q=0.1', 'application/json, text/html; q=0.8'])
    #   # => List['application/json', 'text/html', '*/*']
    #
    module List
      HEADER_DELIMITER    = ','
      PARAMETER_DELIMITER = ';'

      module_function

      def parse(combined, entry_klazz:)
        Array(combined).map { |line| line.split(HEADER_DELIMITER) }.flatten.each_with_index.map do |entry, index|
          value, *parameters = entry.strip.split(PARAMETER_DELIMITER)
          indexed_parameters = Hash[Array(parameters).map { |p| p.strip.split('=') }].transform_keys!(&:to_sym)
          entry_klazz.new(value, index: index, parameters: indexed_parameters)
        end
      end

      def new(combined, entry_klazz:)
        result = parse(combined, entry_klazz: entry_klazz)
        entry_klazz.instance_methods(false).include?(:<=>) ? result.sort! : result
      end

      def to_header(list)
        # noinspection RubyBlockToMethodReference
        list.map { |entry| stringify_entry(entry) }
            .join("#{HEADER_DELIMITER} ")
      end

      def stringify_entry(entry)
        return entry.to_header if entry.respond_to?(:to_header)
        return entry.to_s if entry.respond_to?(:to_s)
        entry.inspect
      end
    end
  end
end
